Proposals/20200511-HVPA-design.md

```diff
@@ -0,0 +1,129 @@
+
+### Goals
+
+1. HVPA should be able to scale an application as per the specifications provided by the users
+
+## Proposal
+
+Users will provide buckets of resources each of which would include minimum and maximum CPU, memory and replicas respectively. A configuration is said to belong to a bucket if total resources allocated to all its replcas lie in that bucket. For eg. if their are 3 replicas with 4 cores each, i.e., total 12 cores, then this configuration could belong to a bucket with:
+```
+minimum CPU: 3
+maximum CPU: 6
+minimum memory: 2.5G
+maximum memory: 7G
+minimum replicas: 2
+maximum replicas: 5
+```
+For this bucket mimimum and maximum total CPU are 6 (= 3 * 2) and 30 (= 6 * 5) respectively, and same for the memory would be 5G and 35G respectively.
+So a configuration with 3 replicas with 6GB memory each would also lie in the same bucket.
+
+HVPA controller would calculate the total memory and CPU recommended by the recommenders (eg. HPA or VPA) by multiplying pod resources and the number of replicas, and would find the bucket for the calculated total resources. HVPA controller would then divide the bucket linearly to calculate the actual `replicas` and `resources` to which it will scale the application such that the total allocated resources to all its replicas is same as recommeded by the recommenders.
+
+### User Stories (optional)
+
+#### Story 1
+A user wants to scale both horizontally and vertically with folloing constraints:
+1. minimum replicas: 1, maximum replicas: 8
+1. minimum memory: 500M, maximum memory: 25G
+1. mimimum CPU: 0m, maximum CPU: 24000m
+
+The above requirement can be denoted as a single bucket.
+HVPA controller would divide the "CPU sub-bucket" and the "memory sub-bucket" linearly among the replica counts such that:
+1. each CPU bucket is of same size, ie, `3000m`
+1. each memory bucket is of same size, ie, `(10G - 100M) / 8`or `1.18G`
+
+Following is an image describing the divided CPU bucket. The numbers in the shaded boxes denote the maximum total resources that can be allocated for a replica count.
+![CPU Story 1](./images/CPUStory1.png)
+
+HVPA controller will take the total cpu and memory resources recommended by HPA or VPA, and will scale the application both horizontally and vertically such that:
+1. the total resoureces allocated to application is same as that recommended by HPA or VPA
+1. the total allocated resource is not more than the maximum total resource for that replica count. This way the resource per replica can be capped to a certain maximum value
+
+For example: If an application is running with 2 replicas with 6 cores each, and VPA and HPA recommendations are 12 cores and 4 replicas respectively, then HVPA would scale the application to 3 replicas and 8 cores. This is because both HPA and VPA recommend 24 cores, and this total lies in the 3rd column of 3 replicas. 
+
```

Comment:
I don't think we talked about that yet and if, then my view is skewed by the API server and its cache where I would like to see vertical scaling before horizontal scaling. So, I am not sure whether this should be the default behaviour and this in the end won't need yet another switch to tweak it. At least, it's controversial.

Here the notes from my file making the proposal to go for vertical scaling first:
- When scaling up, VPA is preferred over HPA (within a bucket; across buckets it’s always HPA by necessity), i.e. only if (1) HPA suggests to increase the replicas and
  (2) this cannot be mitigated anymore by what VPA says, the horizontal recommendation is applied.
  For instance, if HPA wants 1->2, but VPA wants 200m->400m, which is well within the max of 500m, VPA is applied and HPA is not.
  However, if HPA wants 1->2, and VPA wants 400m->600m, then HPA is applied, because VPA is no longer fitting into the current bucket.
- When scaling down, HPA is preferred over VPA (within a bucket; but not across buckets if it can be helped).
  For instance, if HPA wants 4->3, and VPA wants 6000->4000, then HPA is applied and VPA is not.
  However, if HPA wants 3->2, but VPA wants 4000->2000, then VPA is applied, because VPA is still fitting into the current bucket.

(P.S.: Makes only sense with the example, too)

Reply:
These cases are indeed in line with what is described. For behaviour described in your first bullet point a typical bucket could look like:
```yaml
  scaleIntervals:
    - cpu:
        minValue: "0",
        maxValue: "1",
      replicas:
        minValue: "1",
        maxValue: "2",
```
For this particular bucket, minimum CPU = 0 and maximum **aggregated** CPU = 2 (2 replicas of 1000m each).
For scaling calculations, since this is a "linear bucket", the `max - min` CPU per replica (y axis) would be divided **linearly** among the replica counts (number of different counts of replicas for the bucket = maxVValue - minValue + 1 = 2 in this case), which would be 500m.
If **aggregated** CPU recommendation is less than 500m, then we would always have 1 replica, and if the recommendation is more than 500m, only then we would switch to 2 replicas.

The same behaviour can also be achieved using following buckets:
```yaml
  scaleIntervals:
    - cpu:
        minValue: "0",
        maxValue: "500m",
      replicas:
        minValue: "1",
        maxValue: "1",
    - cpu:
        minValue: "250m",
        maxValue: "<anyvalue>",
      replicas:
        minValue: "2",
        maxValue: "<anyvalue>",
```

i hope this clarifies.
/cc @amshuman-kr

Reply:
> These cases are indeed in line with what is described. 

Where is that/what described?